### PR TITLE
Implement Strict Assertion in Unit Testing where needed

### PR DESCRIPTION
**Pull request by oakbani**
🖊 Added [IcecaveStudios/parity](https://github.com/IcecaveStudios/parity) - to perform deep object comparison. MIT Licensed.
🖊  Used SebastianBergmann\Diff\Differ in EventBuilder - which is a part of PHP Unit
🖊 Replaced assertEquals (==) by assertSame(===) for primitive values, where relevant. 
🖊 Modified EventPayloadObject properties sequence to exactly match the testing object.

### CODE REVIEW

File: src/Optimizely/Utils/EventTagUtils.php [hunk 1]
```diff
@@ -43,7 +43,7 @@ public static function getRevenueValue($eventTags) {
             return null;
         }
 
-        if (!isset($eventTags[self::REVENUE_EVENT_METRIC_NAME]) or !$eventTags[self::REVENUE_EVENT_METRIC_NAME]) {
+        if (!isset($eventTags[self::REVENUE_EVENT_METRIC_NAME])) {
```
Comment (oakbani) [inline src/Optimizely/Utils/EventTagUtils.php]: Avoid returning null from here when $eventTags[self::REVENUE_EVENT_METRIC_NAME] is 0 or false

**Comment by msohailhussain** [comment]
@asaschachar can you please assign someone on this PR.

**Comment by asaschachar** [comment]
@msohailhussain Just added @aliabbasrizvi and @alda-optimizely as reviewers

File: src/Optimizely/Bucketer.php [hunk 2]
```diff
@@ -84,7 +84,7 @@ protected function generateBucketValue($bucketingKey)
     {
         $hashCode = $this->generateHashCode($bucketingKey);
         $ratio = $hashCode / Bucketer::$MAX_HASH_VALUE;
-        return floor($ratio * Bucketer::$MAX_TRAFFIC_VALUE);
+        return intval(floor($ratio * Bucketer::$MAX_TRAFFIC_VALUE));
```
Comment (aliabbasrizvi) [inline src/Optimizely/Bucketer.php]: Why is this change needed?

File: src/Optimizely/Event/Builder/EventBuilder.php [hunk 3]
```diff
@@ -169,16 +170,16 @@ private function getConversionParams($config, $eventKey, $experimentVariationMap
                         
             $singleSnapshot[DECISIONS] = [
                 [
-                    CAMPAIGN_ID => $experiment->getLayerId(),
-                    EXPERIMENT_ID => $experimentId,
-                    VARIATION_ID => $variationId
+                    VARIATION_ID => $variationId,
```
Comment (aliabbasrizvi) [inline src/Optimizely/Event/Builder/EventBuilder.php]: The previous ordering was correct.

File: src/Optimizely/Event/Builder/EventBuilder.php [hunk 4]
```diff
@@ -65,19 +65,19 @@ class EventBuilder
     private function getCommonParams($config, $userId, $attributes)
     {
         $visitor = [
-            SNAPSHOTS=> [],
+            ATTRIBUTES => [],
```
Comment (aliabbasrizvi) [inline src/Optimizely/Event/Builder/EventBuilder.php]: The previous ordering was fine.

File: src/Optimizely/Utils/EventTagUtils.php (hunk 1 above)
Comment (aliabbasrizvi) [reply src/Optimizely/Utils/EventTagUtils.php]: Why the change in behavior?

File: tests/BucketerTest.php [hunk 5]
```diff
@@ -55,23 +55,23 @@ public function testGenerateBucketValue()
...
             2434,
             $generateBucketValueMethod->invoke(new Bucketer($this->loggerMock), $this->getBucketingKey('ppid2', '1886780722'))
         );
-        $this->assertEquals(
+        $this->assertSame(
             5439,
             $generateBucketValueMethod->invoke(new Bucketer($this->loggerMock), $this->getBucketingKey('ppid3', '1886780721'))
         );
-        $this->assertEquals(
+        $this->assertSame(
```
Comment (aliabbasrizvi) [inline tests/BucketerTest.php]: Given the value being checked is already primitive, there is no reason to change it

File: tests/EventTests/EventBuilderTest.php [hunk 6]
```diff
@@ -16,13 +16,16 @@
...
+
 use Optimizely\Bucketer;
 use Optimizely\DecisionService\DecisionService;
 use Optimizely\ErrorHandler\NoOpErrorHandler;
 use Optimizely\Event\Builder\EventBuilder;
 use Optimizely\Event\LogEvent;
 use Optimizely\Logger\NoOpLogger;
 use Optimizely\ProjectConfig;
+use Icecave\Parity\Parity;
+use SebastianBergmann\Diff\Differ;
```
Comment (aliabbasrizvi) [inline tests/EventTests/EventBuilderTest.php]: nit. Include external packages at the top before internal packages.

File: tests/EventTests/EventBuilderTest.php [hunk 7]
```diff
@@ -841,6 +873,7 @@ public function testCreateConversionEventWithBucketingIDAttribute()
         );
 
         $logEvent = $this->fakeParamsToReconcile($logEvent);
-        $this->assertEquals($expectedLogEvent, $logEvent);
+        $result = $this->areLogEventsEqual($expectedLogEvent,$logEvent);
+        $this->assertTrue($result[0],$result[1]);
```
Comment (aliabbasrizvi) [inline tests/EventTests/EventBuilderTest.php]: nit. add a space after all comas (,)

File: tests/UtilsTests/EventTagUtilsTest.php [hunk 8]
```diff
@@ -60,9 +60,9 @@ public function testGetRevenueValueWithInvalidRevenueTag() {
     }
 
     public function testGetRevenueValueWithRevenueTag() {
-        $this->assertEquals(65536, EventTagUtils::getRevenueValue(array('revenue' => 65536)));
-        $this->assertEquals(9223372036854775807, EventTagUtils::getRevenueValue(array('revenue' => 9223372036854775807)));
-        $this->assertEquals(0, EventTagUtils::getRevenueValue(array('revenue' => 0)));
+        $this->assertSame(65536, EventTagUtils::getRevenueValue(array('revenue' => 65536)));
+        $this->assertSame(9223372036854775807, EventTagUtils::getRevenueValue(array('revenue' => 9223372036854775807)));
+        $this->assertSame(0, EventTagUtils::getRevenueValue(array('revenue' => 0)));
```
Comment (aliabbasrizvi) [inline tests/UtilsTests/EventTagUtilsTest.php]: Similar comment as earlier. Comparison to a similar value does not need this change.

File: src/Optimizely/Utils/EventTagUtils.php (hunk 1 above)
Comment (oakbani) [reply src/Optimizely/Utils/EventTagUtils.php]: !$eventTags[self::REVENUE_EVENT_METRIC_NAME] 
This line also holds true when the value is 0. Hence, it returns null. This is not our intended behavior.
The condition : (!isset($eventTags[self::REVENUE_EVENT_METRIC_NAME])
suffices to check that the dictionary key exists and it is not null

File: src/Optimizely/Bucketer.php (hunk 2 above)
Comment (oakbani) [reply src/Optimizely/Bucketer.php]: Because the method doc says that it would return an integer.

File: src/Optimizely/Event/Builder/EventBuilder.php (hunk 4 above)
Comment (oakbani) [reply src/Optimizely/Event/Builder/EventBuilder.php]: The ordering isn't significant when sending the payload to event endpoint. I have changed the ordering here to sync it with the test object ordering in the unit test cases. Our new Parity strict checking also check the ordering like it does in Ruby

File: src/Optimizely/Event/Builder/EventBuilder.php (hunk 3 above)
Comment (oakbani) [reply src/Optimizely/Event/Builder/EventBuilder.php]: Please see my above comment

File: tests/BucketerTest.php (hunk 5 above)
Comment (oakbani) [reply tests/BucketerTest.php]: In PHP, 
6128 == "6128" is true
6128 = 6128.0 is true
hence, assertSame should be used to check type as well

Comment (oakbani) [reply tests/BucketerTest.php]: In fact, assertSame can not be used with Object references because === checks that both references point to the same memory location. For deep strict object comparison, I have introduced Parity

File: tests/UtilsTests/EventTagUtilsTest.php (hunk 8 above)
Comment (oakbani) [reply tests/UtilsTests/EventTagUtilsTest.php]: 0 == null, 0 == false, 0 == "0.0" and 0 == 0.0. All are true.
assertSame enforces === . Hence no type conversion. Similar for others

File: src/Optimizely/Event/Builder/EventBuilder.php (hunk 4 above)
Comment (aliabbasrizvi) [reply src/Optimizely/Event/Builder/EventBuilder.php]: The previous ordering was correct because presentability in code is also important. Perhaps the test should be fixed to match it.

File: src/Optimizely/Event/Builder/EventBuilder.php (hunk 3 above)
Comment (aliabbasrizvi) [reply src/Optimizely/Event/Builder/EventBuilder.php]: Same response.

File: tests/BucketerTest.php (hunk 5 above)
Comment (aliabbasrizvi) [reply tests/BucketerTest.php]: Sounds good. Thanks.

File: tests/UtilsTests/EventTagUtilsTest.php (hunk 8 above)
Comment (aliabbasrizvi) [reply tests/UtilsTests/EventTagUtilsTest.php]: Perfect.

**Comment by msohailhussain** [comment]
@aliabbasrizvi request changes are made and updated the PR. Please check https://github.com/optimizely/php-sdk/commit/e60d1c538bbefdbb2415eba8a762de7c991ed5c6

**Comment by msohailhussain** [comment]
@aliabbasrizvi @alda-optimizely  Can anyone of you merge the code, if nothing is pending. PR is already approved.
